numpy/lib/tests/test_arraysetops.py

```diff
@@ -563,6 +563,34 @@ def test_unique_axis(self):
         result = np.array([[-0.0, 0.0]])
         assert_array_equal(unique(data, axis=0), result, msg)
 
+    def test_unique_axis_zeros(self):
+        # issue 15559
+        single_zero = np.empty(shape=(2, 0), dtype=np.int8)
+        uniq, idx, inv, cnt = unique(single_zero, axis=0, return_index=True,
+                                     return_inverse=True, return_counts=True)
+
+        assert_equal(uniq.dtype, single_zero.dtype)
+        msg = "Unique with shape=(2, 0) and axis=0 failed"
+        assert_array_equal(uniq, np.empty(shape=(0, 0)), msg)
```

Comment:
It's possible I'm thinking of `axis=1` and not `axis=0`, but this result is wrong, and should be
```suggestion
        assert_array_equal(uniq, np.empty(shape=(1, 0)), msg)
```
There is exactly _one_ unique empty object, not zero of them.

Reply:
Ah, yeah, you're totally correct. I've rewritten it to not try to reuse the `_unique1d` code, and instead just compute the answers manually. It's a bit unfortunate in my mind that this edge case requires so much extra code, but I can't think of a better way to do it here.

I guess another approach would be to generalise `ndarray.reshape` and `view` to handle empty/zero-sized data better. Do you have thoughts on whether that would be a better way?

Reply:
Unfortunately the semantics of `reshape` and `view` are both ungeneralizable.

It's my opinion that `reshape` is just a bad API - 90% or more of it's use cases are to collapse adjacent axes together, yet it neither is able to check that that actually happens nor perform that action for 0-size arrays.

View would also need an API change, perhaps something like `.view(dt, axis=None)` to indicate that no axes may be resized (the default would be `-1`).

Both are interesting things to discuss further, but I wouldn't attempt to combine them with this patch.

Reply:
Ok, thanks for the info.